pkg/volume/gce_pd/gce_util.go

```diff
@@ -122,9 +137,111 @@ func (util *GCEDiskUtil) DetachDisk(pd *gcePersistentDisk) error {
 	if err := gce.(*gce_cloud.GCECloud).DetachDisk(pd.pdName); err != nil {
 		return err
 	}
+
+	devicePaths := []string{
+		path.Join("/dev/disk/by-id/", "google-"+pd.pdName),
+		path.Join("/dev/disk/by-id/", "scsi-0Google_PersistentDisk_"+pd.pdName),
+	}
+
+	if pd.partition != "" {
+		for i, path := range devicePaths {
+			devicePaths[i] = path + "-part" + pd.partition
+		}
+	}
+	numTries := 0
+	// Wait for the disk device to detach
+	for {
+		pathExists := false
+		for _, path := range devicePaths {
+			udevadmChangeToDrive(path)
+			_, err := os.Stat(path)
+			if err == nil {
+				pathExists = true
+			}
+			if err != nil && !os.IsNotExist(err) {
+				return err
+			}
+		}
+		if !pathExists {
+			break
+		}
+		numTries++
+		if numTries == 10 {
+			return errors.New("Could not detach disk: Timeout after 10s")
+		}
+		time.Sleep(time.Second)
+	}
+
 	return nil
 }
 
+// Calls "udevadm trigger --action=change" for newly created "/dev/sd*" drives (exist only in after set).
+// This is workaround for Issue #7972. Once the underlying issue has been resolved, this may be removed.
+func udevadmChangeToNewDrives(sdBefore, sdAfter []string) {
+	sdexists := make(map[string]bool)
+	for _, sd := range sdBefore {
+		sdexists[sd] = true
+	}
+	for _, sd := range sdAfter {
+		if !sdexists[sd] {
+			udevadmChangeToDrive(sd)
+		}
+	}
+}
+
+// Calls "udevadm trigger --action=change" on the specified drive.
+// drivePath must be the the block device path to trigger on, in the format "/dev/sd*", or a symlink to it.
+// This is workaround for Issue #7972. Once the underlying issue has been resolved, this may be removed.
+func udevadmChangeToDrive(drivePath string) {
+	glog.Infof("udevadmChangeToDrive: drive=%q", drivePath)
+
+	// Evaluate symlink, if any
+	drive, err := filepath.EvalSymlinks(drivePath)
```

Comment:
If a path is not a symlink `evalSymlinks` simply returns the same path, it does not error. For example, `/dev/sda` would just return `/dev/sda`. Since there is no harm in calling this on a non-symlink path and it helps avoid unnecessarily adding logic to decide when it should be called, I'll leave it as is. Let me know if you disagree.